src/vm/functions/iterables.rs

```diff
@@ -0,0 +1,179 @@
+                let v = Value::buff_from(vec![x.clone()])?;
+                let argument = [ SymbolicExpression::atom_value(v) ];
+                let filter_eval = apply(&function, &argument, env, context)?;
+                if let Value::Bool(include) = filter_eval {
+                    if include {
+                        filtered_vec.push(x);
+                    } // else, filter out.
+                } else {
+                    return Err(CheckErrors::TypeValueError(BoolType, filter_eval).into())
+                }
+            }
+            Value::buff_from(filtered_vec)
+        },
+        _ => Err(CheckErrors::ExpectedListOrBuffer(TypeSignature::type_of(&iterable)).into())
+    }
+}
+
+pub fn native_fold(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    check_argument_count(3, args)?;
+
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::ExpectedName)?;
+
+    let function = lookup_function(&function_name, env)?;
+    let iterable = eval(&args[1], env, context)?;
+    let initial = eval(&args[2], env, context)?;
+
+    let mapped_args: Vec<_> = match iterable {
+        Value::List(mut list) => {
+            list.data.drain(..).map(|x| {
+                SymbolicExpression::atom_value(x)
+            }).collect()
+        },
+        Value::Buffer(mut buff) => {
+            buff.data.drain(..).map(|x| {
+                SymbolicExpression::atom_value(Value::buff_from_byte(x))
+            }).collect()
+        },
+        _ => return Err(CheckErrors::ExpectedListOrBuffer(TypeSignature::type_of(&iterable)).into())
+    };
+    mapped_args.iter().try_fold(initial, |acc, x| {
+        apply(&function, &[x.clone(), SymbolicExpression::atom_value(acc)], env, context)
+    })
+}
+
+pub fn native_map(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    check_argument_count(2, args)?;
+
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::ExpectedName)?;
+    let iterable = eval(&args[1], env, context)?;
+    let function = lookup_function(&function_name, env)?;
+
+    let mapped_args: Vec<_> = match iterable {
+        Value::List(mut list) => {
+            list.data.drain(..).map(|x| {
+                vec![SymbolicExpression::atom_value(x)]
+            }).collect()
+        },
+        Value::Buffer(mut buff) => {
+            buff.data.drain(..).map(|x| {
+                vec![SymbolicExpression::atom_value(Value::buff_from_byte(x))]
+            }).collect()
+        },
+        _ => return Err(CheckErrors::ExpectedListOrBuffer(TypeSignature::type_of(&iterable)).into())
+    };
+    let mapped_vec: Result<Vec<_>> =
+        mapped_args.iter().map(|argument| apply(&function, &argument, env, context)).collect();
+    Value::list_from(mapped_vec?)
+}
+
+pub fn native_append(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
```

Comment:
Can I append to a `buff`?

Reply:
you would use `concat`.
I think `append` is really just a nicety so that you can append an element to a list, instead of wrapping it in a list and then concatenating.

Reply:
It might be worth adding `append` for `buff` if it's just a few extra lines to do so?  @kantai what do you think?

Reply:
My previous implementation was supporting the usage of `append` with `buff`. I got rid of this approach after talking with @kantai, the main point being that it's clearer to explain that `concat` should always be used with a buffer, instead of `(append "block" "s")`, but `(concat "block" "stack")` instead of `(append "block" "stack")`, which can only append a `(buff 1)`.

Reply:
Yeah -- I don't think it makes sense to have `append` for buffers, when `concat` is the same function, with a more permissive type signature.

Reply:
Makes sense to me.  This is consistent with a comment I had earlier about how distinct `buff` and `list` will be to programmers.  This sounds to me like they'll be very much distinct?  i.e. we won't try to treat a `buff` as a special list.